File: Logging/LogParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace DiagonalUI.Logging
{
    class Line
    {
        public string Message { get; set; }
        public object Data { get; set; }
    }
    public delegate void LineHandler(object data);
    internal class LogParser
    {
        private readonly Dictionary<string, LineHandler> _handlers = new();

        private readonly string _tag;
        public LogParser(string tag)
        {
            _tag = tag;
        }

        public void AddHandler(string messageName, LineHandler handler)
        {
            _handlers.Add(messageName, handler);
        }

        public void ProcessLine(string line)
        {
            var tagComment = "/* log:tag:" + _tag + " */";
            var index = line.IndexOf(tagComment);
            if (index != -1)
            {
                var messageData = line.Substring(index + tagComment.Length);
                var deserialized = JsonSerializer.Deserialize<Line>(messageData);
                if(deserialized != null && _handlers.ContainsKey(deserialized.Message))
                {
                    _handlers.GetValueOrDefault(deserialized.Message)(deserialized.Data);
                }
            }
        }
    }
}

File: src/main/kotlin/com/rutins/aleks/diagonalui/LogParser.kt
package com.rutins.aleks.diagonalui

import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

class LogParser(private val tag: String) {
    data class Line(val message: String, val data: JsonElement)

    private val handlers = HashMap<String, (JsonElement) -> Unit>()

    fun addHandler(messageName: String, handler: (JsonElement) -> Unit) {
        handlers.put(messageName, handler)
    }

    fun processLine(line: String) {
        val tagComment = "/* log:tag:$tag */"
        val index = line.indexOf(tagComment)
        if(index != -1) {
            val messageData = line.substring(index + tagComment.length)
            try {
                val deserialized = Json.decodeFromString<Line>(messageData)
                if (deserialized.message in handlers) {
                    handlers[deserialized.message]?.invoke(deserialized.data)
                }
            } catch (_: Throwable) {

            }
        }
    }
}